stdlib/test/builtin/test__hash.mojo

```diff
@@ -0,0 +1,803 @@
+        hash[SIMD[DType.uint64, 4](1, 0, 0, 0)]("d".unsafe_ptr(), 1),
+    )
+
+
+def test_avalanche():
+    # test that a input with one bit value produces
+    var data = UnsafePointer[UInt8].alloc(256)
+    memset_zero(data, 256)
+    var hashes0 = List[UInt]()
+    var hashes1 = List[UInt]()
+    hashes0.append(hash(data, 256))
+    hashes1.append(hash[SIMD[DType.uint64, 4](0, 1, 0, 0)](data, 256))
+
+    for i in range(256):
+        memset_zero(data, 256)
+        var v = 1 << (i & 7)
+        data[i >> 3] = v
+        hashes0.append(hash(data, 256))
+        hashes1.append(hash[SIMD[DType.uint64, 4](0, 1, 0, 0)](data, 256))
+
+    for i in range(len(hashes0)):
+        var diff = dif_bits(hashes0[i], hashes1[i])
+        assert_true(
+            diff > 16,
+            str("Index: {}, diff between: {} and {} is: {}").format(
+                i, hashes0[i], hashes1[i], diff
+            ),
+        )
+
+    assert_dif_hashes(hashes0, 14)
+    assert_dif_hashes(hashes1, 15)
+
+
+def test_trailing_zeros():
+    var data = UnsafePointer[UInt8].alloc(8)
+    data[0] = 23
+    var hashes0 = List[UInt]()
+    var hashes1 = List[UInt]()
+    for i in range(1, 9):
+        hashes0.append(hash(data, i))
+        hashes1.append(hash[SIMD[DType.uint64, 4](0, 1, 0, 0)](data, i))
+
+    for i in range(len(hashes0)):
+        var diff = dif_bits(hashes0[i], hashes1[i])
+        assert_true(
+            diff > 19,
+            str("Index: {}, diff between: {} and {} is: {}").format(
+                i, hashes0[i], hashes1[i], diff
+            ),
+        )
+
+    assert_dif_hashes(hashes0, 18)
+    assert_dif_hashes(hashes1, 18)
+
+
+def assert_fill_factor[
+    label: String
+](words: List[String], num_buckets: Int, lower_bound: Float64):
+    # A perfect hash function is when the number of buckets is equal to number of words
+    # and the fill factor results in 1.0
+    var buckets = List[Int](0) * num_buckets
+    for w in words:
+        var h = hash(w[].unsafe_ptr(), w[].byte_length())
+        buckets[h % num_buckets] += 1
+    var unfilled = 0
+    for v in buckets:
+        if v[] == 0:
+            unfilled += 1
+
+    var fill_factor = 1 - unfilled / num_buckets
+    assert_true(
+        fill_factor >= lower_bound,
+        str("Fill factor for {} is {}, provided lower boound was {}").format(
+            label, fill_factor, lower_bound
+        ),
+    )
+
+
+def assert_fill_factor_old_hash[
+    label: String
+](words: List[String], num_buckets: Int, lower_bound: Float64):
+    # A perfect hash function is when the number of buckets is equal to number of words
+    # and the fill factor results in 1.0
+    var buckets = List[Int](0) * num_buckets
+    for w in words:
+        var h = old_hash(w[].unsafe_ptr(), w[].byte_length())
+        buckets[h % num_buckets] += 1
+    var unfilled = 0
+    for v in buckets:
+        if v[] == 0:
+            unfilled += 1
+
+    var fill_factor = 1 - unfilled / num_buckets
+    assert_true(
```

Comment:
FYI I'm seeing this fail internally on a m7i AWS VM with error:

```
stdlib/test/builtin/test__hash.mojo:731:16: AssertionError: Fill factor for AR is 0.59550286040551736, provided lower boound was 0.62
```

Should we adjust the bound, or is there an error elsewhere?

Reply:
Ah yes, the results of old hash are arch dependent (because of difference in SIMD width). I already observed some flotation here in CI and corrected for it. @JoeLoser I will change the value to be `0.59`